pkg/operator/encryption/migration_controller.go

```diff
@@ -198,46 +196,73 @@ func (c *migrationController) migrateKeysIfNeededAndRevisionStable() (resetProgr
 			continue
 		}
 
-		// storage migration takes a long time so we expose that via a distinct status change
-		if err := c.setProgressing(strings.Title(groupToHumanReadable(gr))+strings.Title(gr.Resource), "migrating resource %s.%s to new write key", groupToHumanReadable(gr), gr.Resource); err != nil {
-			return false, err
-		}
-
 		if needsMigration(writeSecret, gr) {
-			migrationErr := c.runStorageMigration(gr)
-			errs = append(errs, migrationErr)
-			if migrationErr != nil {
-				continue
+			// storage migration takes a long time so we expose that via a distinct status change
+			if err := c.setProgressing(strings.Title(groupToHumanReadable(gr))+strings.Title(gr.Resource), "migrating resource %s.%s to new write key", groupToHumanReadable(gr), gr.Resource); err != nil {
+				return false, err
 			}
-		}
 
-		tracker := trackers[writeSecret.Name]
-		tracker.secret = writeSecret
-		tracker.resources = append(tracker.resources, gr)
-		trackers[writeSecret.Name] = tracker
-	}
+			if err := c.runStorageMigration(gr); err != nil {
+				return false, err
+			}
+
+			// update secret annotations
+			if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+				s, err := c.secretClient.Secrets(writeSecret.Namespace).Get(writeSecret.Name, metav1.GetOptions{})
+				if err != nil {
+					return fmt.Errorf("failed to get key secret %s/%s: %v", writeSecret.Namespace, writeSecret.Name, err)
+				}
 
-	for _, tracker := range trackers {
-		migratedSecret := tracker.secret.DeepCopy()
-		if migratedSecret.Annotations == nil {
-			migratedSecret.Annotations = map[string]string{}
+				changed, err := setResourceMigrated(gr, s)
+				if !changed {
+					return nil
+				}
+
+				_, _, updateErr := resourceapply.ApplySecret(c.secretClient, c.eventRecorder, s)
+				return updateErr
+			}); err != nil {
+				return false, err
+			}
 		}
-		migratedSecret.Annotations[encryptionSecretMigratedTimestamp] = time.Now().Format(time.RFC3339)
-		migratedResourceBytes, marshallErr := json.Marshal(migratedGroupResources{Resources: tracker.resources})
-		if marshallErr != nil {
-			errs = append(errs, marshallErr)
-			continue
+	}
+
+	// if we reach this, all migration went fine and we can reset progressing condition
+	return true, nil
+}
+
+func setResourceMigrated(gr schema.GroupResource, s *corev1.Secret) (bool, error) {
+	migratedGRs := migratedGroupResources{}
+	if existing, found := s.Annotations[encryptionSecretMigratedResources]; found {
+		if err := json.Unmarshal([]byte(existing), &migratedGRs); err != nil {
+			// ignore error and just start fresh
```

Comment:
the worst-case scenario is that we will run migration one more time - why not :)